Document offline source code analysis (#225)
tidalmigrations/tidalmigrations.github.io: pages/tidal tools/sourcecode.md

The analysis will look at all of these factors and determine an overall migration difficulty.

Try it out!

{% include image.html file="source_code_analyze.png" caption="Analyze your source code" %}

## Getting Started

> 1. Before you can analyze a source code and rank your applications by *Migration Difficulty*, You must enable the Source Code Analysis feature for your account. To do so, go to your workspace Settings/Preferences (https://yourworkspace.tidalmg.com/#/admin/setting). You will find the Source Code Analysis section at the bottom of the page.

Reviewer: ```suggestion
1. Before you can analyze a source code and rank your applications by *Migration Difficulty*, you must enable the Source Code Analysis feature for your account. To do so, go to your workspace Settings/Preferences (https://yourworkspace.tidalmg.com/#/admin/setting). You will find the Source Code Analysis section at the bottom of the page.
```

> 2. Now that you have activated the Source Code Analyze feature, You need to install, configure and authenticate via Tidal Tools. Make sure you follow these guides.
>   - How to [install](tidal-tools.html#install) Tidal Tools.
>   - Install Tidal Tools [dependencies](tidal-tools.html#dependencies).
>   - Make sure you can [connect](tidal-tools.html#connecting-to-the-api) with your workspace using Tidal Tools.
>   - As a last step, You should run the [tidal doctor](troubleshooting.html#tidal-doctor) command to verify that your environment has been configured properly.

Reviewer: ```suggestion
2. Now that you have activated the Source Code Analysis feature, you need to install, configure and authenticate via Tidal Tools. Make sure you follow these guides.
  - How to [install](tidal-tools.html#install) Tidal Tools.
  - Install Tidal Tools [dependencies](tidal-tools.html#dependencies).
  - Make sure you can [connect](tidal-tools.html#connecting-to-the-api) with your workspace using Tidal Tools.
  - As the last step, You should run the [tidal doctor](troubleshooting.html#tidal-doctor) command to verify that your environment has been configured properly.
```

> 3. You will need the ID of the application for which you are going to perform as source code analysis.  - You can find it in the URL bar when looking at an application. ex. If you are viewing an application in Tidal Migrations, the URL will show https://demo.tidalmg.com/#/apps/111 in this case 111 is the application ID.

Reviewer: ```suggestion
3. You will need the ID of the application for which you are going to perform the source code analysis. You can find it in the URL bar when looking at an application. ex. If you are viewing an application in Tidal Migrations, the URL will show `https://demo.tidalmg.com/#/apps/111` in this case 111 is the application ID.
```

> 4. Lastly, You will need a local copy of the source code for the application.

Reviewer: ```suggestion
4. Lastly, you will need a local copy of the source code for the application.
```

> 5. Perform the Source Code Analysis. There are 2 options to do so. You can find more details in this [section](#perform-the-analysis)

Reviewer: ```suggestion
5. Perform the Source Code Analysis. There are 2 options to do so. You can find more details in this [section](#perform-the-analysis).
```

> That is all, You should be able to see the results in your workspace within seconds.

Reviewer: ```suggestion
That is all. You should be able to see the results in your workspace within seconds.
```

{% include tip.html content="Looking to try it out and don't have any code handy? You can use this sample [schoolbus application](https://github.com/tidalmigrations/schoolbus) by [cloning it from GitHub](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/cloning-a-repository)." %}

## Perform the analysis

> When analyzing a source code (or multiple), You have two options. 

Reviewer: ```suggestion
When analyzing a source code (or multiple), you have two options.
```

1. Perform the source code analysis and upload the result immediately to your workspace. This is Tidal Tools' default behaviour. To do so, all you need to do is run the following command.

    ```bash
    cd /path/to/source-code
    tidal analyze code --app-id 111
    ```

    Alternatively, you can pass the location to the source code.
    ```bash
    tidal analyze code ~/location/to-source-code/ --app-id 111
    ```

2. Perform the source code analysis and upload the results **at a later time**. The following section will explain how to run database analysis in **offline mode**

### Running offline

> There are circumstances in which you need to perform a source code analysis on an environment without or with restricted internet access. In such case, You can perform the analysis, capture its results and at a later stage upload those results to your Tidal Migrations workspace.

Reviewer: ```suggestion
There are circumstances in which you need to perform a source code analysis on an environment without or with restricted internet access. In such case, you can perform the analysis, capture its results and at a later stage upload those results to your Tidal Migrations workspace.
```

These are the steps you need to follow in order to bypass internet access limitations:

> 1. On a Machine with internet access, you need to install, and configure Tidal Tools. 

Reviewer: Should we also mention here that the user's air-gapped machine needs Docker installed? ([Reference](https://guides.tidalmg.com/tidal-tools.html#dependencies))

Author: I think we shouldn't go into it. The second step of Getting Started will cover that.

Author: Let me know what you think @pranavmalaviya. If you still feel we should do it. Let's work on that on another PR. We might have to re-word the steps.

> 2. Package Tidal Tools and its required docker container images into a tar file, This will allow you to move the archive file into your restricted environment. To do so, run the following command.

Reviewer: ```suggestion
2. Package Tidal Tools and its required docker container images into a tar file. This will allow you to move the archive file into your restricted environment. To do so, run the following command.
```

    `tidal backup`

    Once it has finished, you will find (in your current location) a tar file called `tidal-snapshot_DATE.tar`. This is the file you need to transport into your internet restricted environment.

> 3. On the machine that has no internet access. You will now restore Tidal Tools using the following command.

Reviewer: ```suggestion
3. On the machine that has no internet access, you will now restore Tidal Tools using the following command.
```

    `tidal restore tidal-snapshot_DATE.tar`

>     This will load a docker image and all the existing Tidal Tools configurations from the original machine. You can now run the source code analysis without any external network connectivity.:

Reviewer: ```suggestion
    This will load a docker image and all the existing Tidal Tools configurations from the original machine. You can now run the source code analysis without any external network connectivity.
```

    `tidal analyze code [path/to-directory] --offline --output-dir [directory]`

    Note:

>     -  The `--offline` flag, it will indicate to Tidal Tools that the output needs to be stored in a file instead of being uploaded.
>     - The `--output-dir`, will be use to determine where the output will be located.

Reviewer: ```suggestion
    - The `--offline` flag indicates to Tidal Tools that the output needs to be stored in a file instead of being uploaded.
    - The `--output-dir` flag determines where the output will be located.
```

    After the analysis is completed, you will find a zip file called, `result-<DATE>.zip` that can then be transferred into a machine with internet connectivity.

> 4. Back to the machine with internet access. you can now upload your results to your workspace with this command.

Reviewer: ```suggestion
4. Back to the machine with internet access, you can now upload your results to your workspace with this command.
```

    `tidal analyze code --app-id [app_id_for_your_application] --upload [file_name]`

>     You should receive confirmation that the upload has been completed and can navigate to Tidal Migrations to see the results.

Reviewer: I wasn't aware of this. How do they get the confirmation? Email?

Author: Yeah, that is right. You will get an email.
